rfcs/0110-Best-practices-for-testing-and-credentials.md

```diff
@@ -0,0 +1,99 @@
+* Proposed by: @djmitche
+
+# Summary
+
+Taskcluster repositories should pass their tests "out of the box": `git clone`, `yarn`, and `yarn test` should result in a successful test run.
+
+This is simple for test suites that do not require credentials [1].
+Otherwise, we must find a way to give useful test results when no credentials are available, while ensuring that all tests get run before code goes into production.
+
+Briefly, this proposal suggests a uniform approach to skipping tests that require external services by default, but ensuring that all of those tests run in CI wherever possible.
+To minimze the loss of test coverage from skipped tests, it suggests use of mocks for external services and running tests against both mocks and actual services, when possible.
+
+<sub>[1] or otherwise access external services; such services generally require credentials.</sub>
+
+*NOTE*: this RFC only applies to JavaScript libraries and services.
+
+## Motivation
+
+The testing situation for the Taskcluster services and libraries varies widely from repo to repo, so it is difficult to know what to expect when developing a pull request.
+for some repositories, `yarn test` does not work out of the box, leading to confusion and friction for contributors.
+For others, we have had misconfigurations that fail to run all tests on pull requests or pushes to master, allowing deployment of code to production with failing tests.
+
+The goals of this proposal, then, are:
+
+* A consistent approach to testing across Taskcluster repositories, with a common set of supporting tools
+* Repositories that are approachable for new contributors
+* Confidence that production code has passed *all* tests
+
+# Details
+
+The bulk of this proposal is to document new best practices, as described below.
+The final best-practices documents would be developed in concert with utilities to support them (in `taskcluster-lib-testing`, for example), and follow the usual review process.
+
+Once those best practices and libraries are done, the work that would remain is to upgrade all services and libraries to follow them, and to develop new mocks for the remaining un-mocked external services.
+
+## Best Practices (outline)
+
+### Credential Management
+
+Tests should be capable of fetching credentials needed to access external services either from an on-disk file (for local development) or the Taskcluster-Secrets service vi Taskcluster-Proxy (for use in CI).
+It should be easy to tell if a specific kind of credential (for example, pulse or AWS) is available.
+
+#### Getting Credentials
+
+Team members and dedicated contributors will want to know how to set up a full set of credentials for themselves.
+Include a section in the README describing how to find or generate credentials for each service.
+
+For Taskcluster credentials, put the necessary scopes in `scopes.txt` and suggest that the user use taskcluster-cli to generate the credentials:
```

Comment:
If we establish this convention, we should have something in `taskcluster-lib-testing` or elsewhere which reads that file and generates temporary credentials with exactly that subset of scopes to ensure that there cannot be drift between what we say you need to pass, and what actually works.  Even if we do the `taskcluster-cli signin` thing from below, the tests should enforce that the only scopes in the credentials are those in this file.

Reply:
I like this idea and I share the concern of scope drift .. but I'm not sure how to implement it.  Calling `auth.currentScopes` with a set of credentials will return the expanded version of those scopes, which may be different if there are any `assume:..` in `scopes.txt`.  When run in CI, that's the only option available since the credentials are not directly accessible from the task.

Perhaps a simpler option is to make a convention of test scopes being encapsulated in a single role, say `project:taskcluster:tests:<projectName>`, and using `--scopes assume:project:taskcluster:tests:<projectName>` in the tc-signin command line.  Then it's simple to visually verify in `.taskcluster.yml` that `task.scopes` only contains that role, and it's simple to modify the role when additional scopes are necessary.  @jonasfj, @walac what do you think?

Reply:
I like the idea of putting all scopes needed for tests in a single role.